<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Imports" data-toc-modified-id="Imports-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Imports</a></span></li><li><span><a href="#%time-vs-%timeit" data-toc-modified-id="%time-vs-%timeit-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>%time vs %timeit</a></span></li><li><span><a href="#*args-und-**kwargs" data-toc-modified-id="*args-und-**kwargs-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>*args und **kwargs</a></span><ul class="toc-item"><li><span><a href="#Beispiel-*args" data-toc-modified-id="Beispiel-*args-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Beispiel *args</a></span></li><li><span><a href="#Beispiel-**kwargs" data-toc-modified-id="Beispiel-**kwargs-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Beispiel **kwargs</a></span></li></ul></li><li><span><a href="#Comprehensions" data-toc-modified-id="Comprehensions-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Comprehensions</a></span><ul class="toc-item"><li><span><a href="#Dictionary" data-toc-modified-id="Dictionary-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Dictionary</a></span></li><li><span><a href="#List" data-toc-modified-id="List-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>List</a></span></li><li><span><a href="#Set" data-toc-modified-id="Set-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Set</a></span></li></ul></li><li><span><a href="#Counter-Funktion" data-toc-modified-id="Counter-Funktion-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Counter Funktion</a></span></li><li><span><a href="#Dataclasses" data-toc-modified-id="Dataclasses-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Dataclasses</a></span><ul class="toc-item"><li><span><a href="#Alt-vs-neu" data-toc-modified-id="Alt-vs-neu-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Alt vs neu</a></span></li></ul></li><li><span><a href="#Docstring" data-toc-modified-id="Docstring-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Docstring</a></span></li><li><span><a href="#Lambda" data-toc-modified-id="Lambda-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Lambda</a></span><ul class="toc-item"><li><span><a href="#map()-Funktion" data-toc-modified-id="map()-Funktion-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>map()-Funktion</a></span></li><li><span><a href="#Sortieren" data-toc-modified-id="Sortieren-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>Sortieren</a></span><ul class="toc-item"><li><span><a href="#Tupel" data-toc-modified-id="Tupel-8.2.1"><span class="toc-item-num">8.2.1&nbsp;&nbsp;</span>Tupel</a></span></li><li><span><a href="#Dictionary" data-toc-modified-id="Dictionary-8.2.2"><span class="toc-item-num">8.2.2&nbsp;&nbsp;</span>Dictionary</a></span></li></ul></li><li><span><a href="#Filtern" data-toc-modified-id="Filtern-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>Filtern</a></span><ul class="toc-item"><li><span><a href="#Liste-mit-int-Werten" data-toc-modified-id="Liste-mit-int-Werten-8.3.1"><span class="toc-item-num">8.3.1&nbsp;&nbsp;</span>Liste mit int-Werten</a></span></li></ul></li></ul></li><li><span><a href="#Mutable-vs-Immutable" data-toc-modified-id="Mutable-vs-Immutable-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Mutable vs Immutable</a></span><ul class="toc-item"><li><span><a href="#Mutable-Beispiel" data-toc-modified-id="Mutable-Beispiel-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Mutable Beispiel</a></span></li><li><span><a href="#Immutable-Beispiel" data-toc-modified-id="Immutable-Beispiel-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Immutable Beispiel</a></span></li><li><span><a href="#Immutable-+-Mutable-Beispiele" data-toc-modified-id="Immutable-+-Mutable-Beispiele-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>Immutable + Mutable Beispiele</a></span><ul class="toc-item"><li><span><a href="#Immutable-mit-Mutable" data-toc-modified-id="Immutable-mit-Mutable-9.3.1"><span class="toc-item-num">9.3.1&nbsp;&nbsp;</span>Immutable mit Mutable</a></span></li><li><span><a href="#Mutable-mit-Immutable" data-toc-modified-id="Mutable-mit-Immutable-9.3.2"><span class="toc-item-num">9.3.2&nbsp;&nbsp;</span>Mutable mit Immutable</a></span></li></ul></li></ul></li><li><span><a href="#OS" data-toc-modified-id="OS-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>OS</a></span><ul class="toc-item"><li><span><a href="#Pfadarten" data-toc-modified-id="Pfadarten-10.1"><span class="toc-item-num">10.1&nbsp;&nbsp;</span>Pfadarten</a></span></li><li><span><a href="#Verzeichnis-erstellen" data-toc-modified-id="Verzeichnis-erstellen-10.2"><span class="toc-item-num">10.2&nbsp;&nbsp;</span>Verzeichnis erstellen</a></span></li><li><span><a href="#Lesen-von-einer-Datei" data-toc-modified-id="Lesen-von-einer-Datei-10.3"><span class="toc-item-num">10.3&nbsp;&nbsp;</span>Lesen von einer Datei</a></span></li><li><span><a href="#Dateien-automatisch-lesen-und-schließen-(with)" data-toc-modified-id="Dateien-automatisch-lesen-und-schließen-(with)-10.4"><span class="toc-item-num">10.4&nbsp;&nbsp;</span>Dateien automatisch lesen und schließen (with)</a></span></li><li><span><a href="#Datei-Zeile-für-Zeile-lesen" data-toc-modified-id="Datei-Zeile-für-Zeile-lesen-10.5"><span class="toc-item-num">10.5&nbsp;&nbsp;</span>Datei Zeile für Zeile lesen</a></span></li><li><span><a href="#Verarbeitung-von-Daten-aus-Dateien" data-toc-modified-id="Verarbeitung-von-Daten-aus-Dateien-10.6"><span class="toc-item-num">10.6&nbsp;&nbsp;</span>Verarbeitung von Daten aus Dateien</a></span></li></ul></li><li><span><a href="#Itertools" data-toc-modified-id="Itertools-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Itertools</a></span><ul class="toc-item"><li><span><a href="#Product" data-toc-modified-id="Product-11.1"><span class="toc-item-num">11.1&nbsp;&nbsp;</span>Product</a></span></li><li><span><a href="#Permutation" data-toc-modified-id="Permutation-11.2"><span class="toc-item-num">11.2&nbsp;&nbsp;</span>Permutation</a></span></li><li><span><a href="#Combinations" data-toc-modified-id="Combinations-11.3"><span class="toc-item-num">11.3&nbsp;&nbsp;</span>Combinations</a></span></li><li><span><a href="#Accumulate" data-toc-modified-id="Accumulate-11.4"><span class="toc-item-num">11.4&nbsp;&nbsp;</span>Accumulate</a></span></li><li><span><a href="#Groupby" data-toc-modified-id="Groupby-11.5"><span class="toc-item-num">11.5&nbsp;&nbsp;</span>Groupby</a></span></li></ul></li></ul></div>

!!! TO-DO !!!

- [ ] Lambda/map()-Funktion - Video zu Lambda weiter schauen (8:37 , https://www.youtube.com/watch?v=Ob9rY6PQMfI)
- [ ] Comprehension/Set noch machen - Video (https://www.youtube.com/watch?v=3dt4OGnU5sM)

# Imports

In [8]:
%run Imports.ipynb

All good


# %time vs %timeit

In [9]:
%timeit sum(range(100000))


3.49 ms ± 374 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [10]:
%time sum(range(100000))

CPU times: total: 15.6 ms
Wall time: 4 ms


4999950000

# &ast;args und &ast;&ast;kwargs

In Python kann eine variable Anzahl von Argumenten an eine Funktion übergeben werdeb, indem man spezielle Symbole verwenden. Es gibt zwei spezielle Symbole:<br>
<br>
<dl>
    <dt>- &ast;args <b>(Argument)</b>  (Nicht-Schlüsselwort-Argumente)</dt>
    <dt>- &ast;&ast;kwargs <b>(Keyword argument)</b> (Schlüsselwort-Argumente)</dt>
</dl>
&ast;args und &ast;&ast;kwargs werden verwendet als Argument/Parameter, wenn man sich nicht sicher ist, wie viele Argumente/Parameter an die Funktion übergeben werden sollen.


## Beispiel &#42;args

In [11]:
def adder(x,y,z):
    print("sum:",x+y+z)

adder(5,10,15,20,35)

# Fehlermeldung wenn zu viele Argumente übergeben werden

TypeError: adder() takes 3 positional arguments but 5 were given

In [ ]:
def adder(*num):
    sum = 0
    
    for n in num:
        sum = sum + n

    print("Sum:",sum)
    
adder(3,4)
adder(10,10,10,10)

Sum: 7
Sum: 40


## Beispiel &#42;&#42;kwargs

In [12]:
def intro(**data):
    print("\nData type of argument:",type(data))

    for key, value in data.items():
        print("{0} is {1}".format(key,value))

intro(Firstname="Sabrina", Lastname="Mayer", Age=22, Phone=1234567890)
intro(Firstname="Max", Lastname="Mueller", Email="johnwood@nomail.com", Country="USA", Age=25, Phone=9876543210)


Data type of argument: <class 'dict'>
Firstname is Sabrina
Lastname is Mayer
Age is 22
Phone is 1234567890

Data type of argument: <class 'dict'>
Firstname is Max
Lastname is Mueller
Email is johnwood@nomail.com
Country is USA
Age is 25
Phone is 9876543210


# Comprehensions

Bietet eine kürzere und verständliche Syntax um verschiedene Formen von Listen zu erstellen

## Dictionary

<b>Beispiel</b>
<br>
Ein {"name":"heros"} für jeden Namen/Helden in zip(names,heros)

In [13]:
names = ["Bruce", "Clark", "Peter", "Logan", "Wade"]
heros = ["Batman", "Superman", "Spiderman", "Wolverine", "Deadpool"]
print(dict(zip(names,heros)))

{'Bruce': 'Batman', 'Clark': 'Superman', 'Peter': 'Spiderman', 'Logan': 'Wolverine', 'Wade': 'Deadpool'}


In [14]:
# Ohne Comprehension
my_dict = {}
for name,hero in zip(names,heros):
    my_dict[name] = hero
print(my_dict)

{'Bruce': 'Batman', 'Clark': 'Superman', 'Peter': 'Spiderman', 'Logan': 'Wolverine', 'Wade': 'Deadpool'}


In [15]:
# Mit Comprehension
my_dict_2 = {name_2:hero_2 for name_2, hero_2 in zip(names,heros)}
print(my_dict_2)

{'Bruce': 'Batman', 'Clark': 'Superman', 'Peter': 'Spiderman', 'Logan': 'Wolverine', 'Wade': 'Deadpool'}


## List

In [16]:
nums = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

<b>Beispiel:</b>
<br>
Jedes Element in nums einer neuen Liste zuweisen

In [17]:
# Ohne Comprehension
my_list = []
for n in nums:
    my_list.append(n)
print(my_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [18]:
# Mit Comprehension
my_list = [n for n in nums]
print(my_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


<b>Beispiel:</b>
<br>
Quadrat von jedem Element in nums in eine neue Liste schreiben

In [19]:
# Ohne Comprehension
my_list = []
for n in nums:
    my_list.append(n*n)
print(my_list)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [20]:
# Mit Comprehension
my_list = [n*n for n in nums]
print(my_list)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [21]:
# Alternative mit lambda + map

my_list = map(lambda x: x*x, nums)
print(list(my_list))

# Muss in eine Liste gemacht werden (Gibt ein Iterator zurück)
# Versionsabhängig (Python2 Liste, Python3 Iterator)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


<b>Beispiel:</b>
<br>
Jedes Element was durch zwei teilbar ist in eine neue Liste schreiben

In [22]:
# Ohne Comprehension
my_list = []
for n in nums:
    if n%2 == 0:
        my_list.append(n)
print(my_list)

[2, 4, 6, 8, 10]


In [23]:
# Mit Comprehension
my_list = [n for n in nums if n%2 == 0]
print(my_list)

[2, 4, 6, 8, 10]


In [24]:
# ALternative mit filter + lambda
my_list = filter(lambda x: x%2 == 0, nums)
print(list(my_list))

[2, 4, 6, 8, 10]


<b>Beispiel</b>
<br>
Ein (Buchstaben, Zahlen) Paar für jeden Buchstaben in "abcd" und jeder Zahl in "0123"

In [25]:
# Ohne Comprehension
my_list = []
for letter in "abcd":
    for number in "0123":
        my_list.append((letter,number))
print(my_list)

[('a', '0'), ('a', '1'), ('a', '2'), ('a', '3'), ('b', '0'), ('b', '1'), ('b', '2'), ('b', '3'), ('c', '0'), ('c', '1'), ('c', '2'), ('c', '3'), ('d', '0'), ('d', '1'), ('d', '2'), ('d', '3')]


In [26]:
# Mit Comprehension
my_list = [(letter,number) for letter in "abcd" for number in "0123"]
print(my_list)

[('a', '0'), ('a', '1'), ('a', '2'), ('a', '3'), ('b', '0'), ('b', '1'), ('b', '2'), ('b', '3'), ('c', '0'), ('c', '1'), ('c', '2'), ('c', '3'), ('d', '0'), ('d', '1'), ('d', '2'), ('d', '3')]


## Set

<b>Beispiel</b>
<br>
Jede Zahl aus der Liste in das Set kopieren

In [27]:
# Ohne Comprehension
nums = [1, 1, 2, 4, 5, 6, 6, 7, 8, 9, 10 ,10]
my_set = set()
for n in nums:
    my_set.add(n)
print(my_set)

{1, 2, 4, 5, 6, 7, 8, 9, 10}


In [28]:
# Mit Comprehension
my_set_2 = {n for n in nums}
print(my_set_2)

{1, 2, 4, 5, 6, 7, 8, 9, 10}


# Counter Funktion

In [29]:
from collections import Counter

In [30]:
x_list = [2,2,3,4,5,5,5,5,5,6,7,3,1,22,22,22,22,22]
count_dict = dict().fromkeys(x_list,0)
for element in x_list:
    count_dict[element] += 1

print(count_dict)

{2: 2, 3: 2, 4: 1, 5: 5, 6: 1, 7: 1, 1: 1, 22: 5}


In [31]:
x = [2,2,3,4,5,5,5,5,5,6,7,3,1,22,22,22,22,22,22]
my_list = Counter(x)
# print(len(my_list.items()))
list_len = len(my_list.items())
if(list_len >= 2):
    most_com_list = my_list.most_common(2)
    print(f"Zahl mit den meisten Einträge: {most_com_list[0][0]}")
    print(f"Anzahl der Einträge: {most_com_list[0][1]}")
    print(f"Es besitzt {list_len} Item(s)")
else:
    print(f"WARNUNG!: Es besitzt nur {list_len} Item(s)")

Zahl mit den meisten Einträge: 22
Anzahl der Einträge: 6
Es besitzt 8 Item(s)


# Dataclasses 

!!! Feature von **Python 3.7** !!!

Wichtig:
- Dataclasses verletzen das Konzept von einer Klassenvariable zum repräsentieren von Instanzvariablen
- Wichtig den @dataclass Dekorator vor die Definition der Klasse schreiben, da man ansonsten Klassenvariablen anstelle von Instanzvariablen erzeugt (Kann zu diversen Problemen führen) 

In [32]:
!python -V

Python 3.10.4


In [33]:
from dataclasses import dataclass, field
import inspect # Funktionen die erzeugt werden ausgeben
from pprint import pprint

## Alt vs neu 

In [34]:
# Alt

class PersonAlt:
    def __init__(self, name:str, address:str):
        self.name = name
        self.address = address
    
    def __str__(self) -> str:
        return f"{self.name}, {self.address}"
    
p1_alt = PersonAlt("Max" , "Am Weiher 4")

print(p1_alt)

pprint(inspect.getmembers(PersonAlt, inspect.isfunction)) # Funktionen der Klasse ausgeben

Max, Am Weiher 4
[('__init__', <function PersonAlt.__init__ at 0x000001F19A40AB00>),
 ('__str__', <function PersonAlt.__str__ at 0x000001F19A40A8C0>)]


**Problem bei Alt:**
- Beim Hinzufügen weiterer Variablen wird der Code unübersichtlicher und ist dementsprechend fehleranfälliger 

In [35]:
# Neu

@dataclass
class PersonNeu:
    name: str
    address: str
        
p2_neu = PersonNeu("Max", "Am Weiher 4")

print(p2_neu)
# die __repr__ Methode wird durch dataclass generiert

pprint(inspect.getmembers(PersonNeu, inspect.isfunction))

PersonNeu(name='Max', address='Am Weiher 4')
[('__eq__', <function PersonNeu.__eq__ at 0x000001F19A40A050>),
 ('__init__', <function PersonNeu.__init__ at 0x000001F19A40AB90>),
 ('__repr__', <function PersonNeu.__repr__ at 0x000001F19A40AEF0>)]


## Default Werte zuweisen

In [36]:
@dataclass
class PersonNeuDef:
    name: str
    address: str
    active: bool = True
    
    # email_addresses: list[str] = [] 
    # Nicht so! Python evaluiert die Standardwerte, wenn es das Skript interpretiert.
    # Folglich würde es bei mehreren Instanzen auf das gleiche Objekt referenzieren. 
    
    email_addresses: list[str] = field(default_factory=list)

p_def_1 = PersonNeuDef("Marius" , "Lembergstraße 5", False)

p_def_2 = PersonNeuDef("Simon", "Am Weiher 2", True, "simon_weiher@gmail.com")
# Die "default_factory" bei "p_def_2" wird nicht aufgerufen, da es einen Wert besitzt.

print(f"{p_def_1}\n{p_def_2}")   

PersonNeuDef(name='Marius', address='Lembergstraße 5', active=False, email_addresses=[])
PersonNeuDef(name='Simon', address='Am Weiher 2', active=True, email_addresses='simon_weiher@gmail.com')


## Ausschließen von Argumenten beim Initialisieren

In [37]:
import random
import string

def gen_id() -> str:
    return "".join(random.choices(string.ascii_uppercase, k=12))

@dataclass
class PersonNeuExcldArgs:
    name: str
    address: str
    active: bool = True
    email_addresses: list[str] = field(default_factory=list)
    
    strength: int = field(repr=False, default=200)
    # repr=False, bedeutet dass es nicht Teil von der __repr__ Ausgabe sein soll.
    
    id: str = field(init=False, default_factory=gen_id)
    # init=False, bedeutet dass diese Instanzvariable nicht Teil der sein Initialisierung wird.

p_excld_args_1 = PersonNeuExcldArgs("Max", "Boschstraße 123", True, "max_bosch@gmail.com")
print(p_excld_args_1)
    
p_excld_args_2 = PersonNeuExcldArgs("Simon", "Am Weiher 2", True, "simon_weiher@gmail.com", id="MCGUAS")
# Führt zu einer Fehlermeldung, da man nicht mehr die id übergeben kann



    
    

PersonNeuExcldArgs(name='Max', address='Boschstraße 123', active=True, email_addresses='max_bosch@gmail.com', id='EVHVXLPHBQBP')


TypeError: PersonNeuExcldArgs.__init__() got an unexpected keyword argument 'id'

## Erzeugen von extra Felder

Mit hilfe von **\_\_post_init\_\_** kann man Werte von anderen Instanzvariablen erzeugen.

In [38]:
import random
import string

def gen_id() -> str:
    return "".join(random.choices(string.ascii_uppercase, k=12))

@dataclass
class PersonNeuPstInit:
    name: str
    address: str
    active: bool = True
    email_addresses: list[str] = field(default_factory=list)
    id: str = field(init=False, default_factory=gen_id)
    search_string: str = field(init=False)
    
    def __post_init__(self) -> (None):
        self.search_string = f"{self.name}, {self.address}"

p_pst_init_1 = PersonNeuPstInit("Max", "Boschstraße 123", True, "max_bosch@gmail.com")
print(p_pst_init_1)

PersonNeuPstInit(name='Max', address='Boschstraße 123', active=True, email_addresses='max_bosch@gmail.com', id='GUQUWVIYWAIH', search_string='Max, Boschstraße 123')


## Freezing

In [39]:
@dataclass(frozen=False) # Standardmäßig ist frozen=False
class PersonNeuNotFrz:
    name: str
    address: str
    active: bool = True
    
p_not_frz_1 = PersonNeuNotFrz("Robert", "Winzerstrasse 5", "False")
print(p_not_frz_1)
p_not_frz_1.name = "Marcel"
print(p_not_frz_1)

###################

@dataclass(frozen=True)
class PersonNeuFrz:
    name: str
    address: str
    active: bool = True
    
p_frz_2 = PersonNeuFrz("Robert", "Winzerstrasse 5", "False")
p_frz_2.name = "Marcel"

PersonNeuNotFrz(name='Robert', address='Winzerstrasse 5', active='False')
PersonNeuNotFrz(name='Marcel', address='Winzerstrasse 5', active='False')


FrozenInstanceError: cannot assign to field 'name'

## Vergleichsoperationen

Mit Hilfe von **order=True** können Vergleichsoperationen zwischen Objekten durchgeführt werden.

In [40]:
@dataclass(order=True)
class PersonNewSlry:
    sort_index: int = field(init=False)
    salary: int
    name: str = field(compare=False)
    address: str = field(compare=False)
    
    def __post_init__(self):
        self.sort_index = self.salary
    
        
p_slry_1 = PersonNewSlry(75000, "Max" , "Main St 123")
p_slry_2 = PersonNewSlry(75000, "Jules", "Main St 456")


print("== : ", p_slry_1 == p_slry_2)
print("<= : ", p_slry_1 <= p_slry_2)
print("<  : ", p_slry_1 < p_slry_2)
print(">= : ", p_slry_1 >= p_slry_2)
print(">  : ", p_slry_1 > p_slry_2)

== :  True
<= :  True
<  :  False
>= :  True
>  :  False


## 3.10 Features

Die folgenden Features sind ab Python Version **3.10** möglich.

In [41]:
!python -V

Python 3.10.4


### kw_only

In [42]:
@dataclass(kw_only=True) # Standardmäßig False
class PersonKwOnly:
    name: str
    address: str
    active: bool = True
    email_addresses: list[str] = field(default_factory=list)

p_kw_only_1 = PersonKwOnly(name="Marius", address="Main St 123")
print(p_kw_only_1)
    
p_kw_only_2 = PersonKwOnly("Marius", "Main St 123") # Fehlermeldung TypeError entsteht, da die Keywords fehlen (in dem Fall "name" und "address")


PersonKwOnly(name='Marius', address='Main St 123', active=True, email_addresses=[])


TypeError: PersonKwOnly.__init__() takes 1 positional argument but 3 were given

### slots

Unter der Haube sind Python Klassen Instanzvariablen in Dictionairies abgespeichert. Hierfür kann die **\_\_dict\_\_** Methode verwendet werden um auf Instanzvariablen zugreifen zu können. <br>
Mit Hilfe von **slots** ist der Zugriff schneller, im Vergleich zu Dictionairies.

In [43]:
# Zugriff mit __dict__
@dataclass
class PersonSlts:
    name: str
    address: str
    active: bool = True
    email_addresses: list[str] = field(default_factory=list)

p_slots = PersonSlts("Max", "Main St 123")
print(p_slots)
print(p_slots.__dict__)
print(p_slots.__dict__["name"])

PersonSlts(name='Max', address='Main St 123', active=True, email_addresses=[])
{'name': 'Max', 'address': 'Main St 123', 'active': True, 'email_addresses': []}
Max


In [44]:
# Vergleich Schnelligkeit (__dict__ vs slots)
import timeit
from functools import partial

@dataclass(slots=False)
class PersonSlotsFalse:
    name: str
    address: str
    email: str

@dataclass(slots=True)    
class PersonSlotsTrue:
    name: str
    address: str
    email: str
    
def get_set_del(person: PersonSlotsFalse | PersonSlotsTrue):
    person.address = "Main St 123"
    person.address
    del person.address

p_slots_f = PersonSlotsFalse("John", "Main St 123", "john_main@gmail.com")
p_slots_t = PersonSlotsTrue("Max", "Second St 456", "max_second@gmail.com")

no_slots = min(timeit.repeat(partial(get_set_del, p_slots_f), number=100_000))
slots = min(timeit.repeat(partial(get_set_del, p_slots_t), number=100_000))

print(f"No slots: {no_slots}")
print(f"Slots: {slots}")

print(f"% Perfomance improvement {(no_slots - slots) / no_slots:.2%}")

No slots: 0.024716999963857234
Slots: 0.019677500007674098
% Perfomance improvement 20.39%


In [45]:
@dataclass(slots=True)    
class PersonSlots:
    name: str
    address: str
    email: str
    
@dataclass(slots=True)
class EmployeeSlots:
    salary: int

class PersonEmployee(PersonSlots, EmployeeSlots):
    pass

p_slots = PersonSlots("Max", "Main St 123", "max_main@gmail.com")

# Problem bei Slots ist, dass es bei mehrfacher Verebung nicht weiß welche Klasse es benutzen soll.

TypeError: multiple bases have instance lay-out conflict

# Docstring

In [46]:
def test():
    """
    Dies ist ein Test:
    
    Hier könnte ihre Werbung stehen
    """

In [47]:
help(test)

Help on function test in module __main__:

test()
    Dies ist ein Test:
    
    Hier könnte ihre Werbung stehen



# Lambda

Aufbau:<br>
<br>
<b>lambda arguments: expression</b>
<br>
<br>
Wird häufig verwendet zum sortieren oder filtern

In [48]:
add_num5 = lambda x: x+5
print(add_num5(5))

10


In [49]:
square_num5 = lambda x: x*5
print(square_num5(5))

25


In [50]:
get_tens = lambda x: int(x/10)%10
print(get_tens(749)) # Gibt die 4 bei 749 aus
print(get_tens(92829.4))

4
2


## map()-Funktion

Die map()-Funktion sorgt dafür, das die Lambda-Funktion auf jedes Element angewendet wird

In [51]:
ls1 = [1,2,3,4,5,6]
ls2 = list(map(lambda x: x ** 2, ls1))
print(ls2)

[1, 4, 9, 16, 25, 36]


## Sortieren

### Tupel

In [52]:
list1 = [("Eier",3.19), ("Honig", 1.89), ("Chips", 2.29),("Red Bull", 1.19)]
list1.sort(key= lambda x: x[0])
print(list1)

[('Chips', 2.29), ('Eier', 3.19), ('Honig', 1.89), ('Red Bull', 1.19)]


### Dictionary

In [53]:
list2 = [{"brand":"Ford", "model":"Focus", "year":2010},{"brand":"Fiat", "model":"Punto", "year":2007},{"brand":"Nissan", "model":"GT-R", "year":2009}]

In [55]:
list2_sorted = sorted(list2, key= lambda x: x["model"])
pprint(list2_sorted)

[{'brand': 'Ford', 'model': 'Focus', 'year': 2010},
 {'brand': 'Nissan', 'model': 'GT-R', 'year': 2009},
 {'brand': 'Fiat', 'model': 'Punto', 'year': 2007}]


## Filtern

### Liste mit int-Werten

In [56]:
list1 = [1,2,3,4,5,6]
list1_filtered = list(filter(lambda x:x%2 == 0, list1))
print(list1_filtered)

[2, 4, 6]


In [58]:
odds = lambda x: x%2 == 1
list1 = [1,2,3,4,5,6]
list1_filtered = list(filter(odds, list1))
print(list1_filtered)

[1, 3, 5]


# Mutable vs Immutable

<b>MUTABLE OBJECTS</b><br>
<dl>
    <dt>- Lists</dt>
    <dt>- Sets</dt>
    <dt>- Dictionaries</dt>
    <dt>- Benutzerdefinierte Klassen</dt>
</dl>
<br>
<b>IMMUTABLE OBJECTS </b>
<dl>
    <dt>- Numbers (Integer, Float, Decimal, Booleans, ...)</dt>
    <dt>- Strings</dt>
    <dt>- Tuples</dt>
    <dt>- Frozen Sets</dt>
    <dt>- Benutzerdefinierte Klassen</dt>
</dl>

## Mutable Beispiel

In [59]:
staedte = ["Berlin", "Paris", "Tokyo"]
for stadt in staedte:
    print(stadt, end=",")

Berlin,Paris,Tokyo,

In [60]:
# Position des erstellten Objekts im Speicher in hexadezimaler Form anzeigen 

print(hex(id(staedte)))

0x1f19a4d9640


In [61]:
# Hinzufügen einer weiteren Stadt

staedte.append("Moskau")
for stadt in staedte:
    print(stadt, end=",")

Berlin,Paris,Tokyo,Moskau,

In [62]:
print(hex(id(staedte)))

0x1f19a4d9640


Das obige Beispiel zeigt, dass man in der Lage ist, den internen Zustand des Objekts "Städte" zu ändern, indem man eine weitere Stadt "Moskau" hinzugefügt hat. Die Speicheradresse des Objekts hat sich nicht geändert.<br>
<br>
Dies bestätigt, dass kein neues Objekt erstellt wurde, sondern dass dasselbe Objekt verändert oder mutiert wurde. Daher kann man sagen, dass das Objekt, das ein Listentyp mit dem Namen der Referenzvariablen 'staedte' ist, ein <b>MUTABLE OBJECT</b> ist.

## Immutable Beispiel

In [63]:
foo = (1,2) # Tupel erzeugen

In [65]:
foo[0] = 3

# Ein Objekt, dessen interner Zustand nicht verändert werden kann, ist <b>IMMUTABLE</b>

TypeError: 'tuple' object does not support item assignment

In [66]:
weekdays = "Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"
print(weekdays)

('Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday')


In [67]:
print(hex(id(weekdays)))

0x1f19a599b40


In [68]:
weekdays += ("Testday",)
print(weekdays)

('Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Testday')


Da Tupel unveränderlich sind, können keine neuen Elemente hinzugefügt werden. Daher wird durch die Zusammenführung von Tupeln mit dem Operator "+" ein neuer imaginärer Tag in das Tupel 'weekdays' eingefügt.

In [69]:
print(hex(id(weekdays)))

0x1f19a5d01c0


Das obige Beispiel mit den Wochentagen zeigt, dass man in der Lage ist, denselben Variablennamen zu verwenden, der auf ein Objekt verweist, das eine Art Tupel mit sieben Elementen ist. Die ID oder der Speicherort des alten und neuen Tupels ist jedoch nicht identisch. <br>
<br>
Man ist nicht in der Lage, den internen Zustand des Objekts "weekdays" zu ändern. Python legt ein neues Objekt in der Speicheradresse an, und die Variable "weekdays" hat begonnen, auf das neue Objekt mit acht Elementen zu verweisen.<br>
Daher kann man sagen, dass das Objekt vom Typ Tupel mit der Referenzvariablen 'weekdays' ein IMMUTABLE OBJEKT ist.

## Immutable + Mutable Beispiele

### Immutable mit Mutable

In [70]:
person = (['Sebastian', 5, 'Maennlich'], ['Sandra', 8, 'Weiblich'])

# Erstelle eine Tupel (Immutable) mit zwei Listen (mutable) als Elemente

In [71]:
print(person)

(['Sebastian', 5, 'Maennlich'], ['Sandra', 8, 'Weiblich'])


In [72]:
print(hex(id(person)))

0x1f19a48b300


In [73]:
person[0][1] = 4

# Veränderung des Alters beim ersten Element

In [74]:
print(hex(id(person)))

0x1f19a48b300


Im obigen Beispiel sieht man, dass das Objekt "person" unveränderlich (IMMUTABLE) ist, da es sich um ein Tupel handelt. Es hat jedoch zwei Listen als Elemente, und wir können den Zustand von Listen ändern (Listen sind veränderbar (MUTABLE)). In diesem Fall haben wir also nicht den Objektverweis innerhalb des Tupels geändert, sondern das referenzierte Objekt wurde geändert.

### Mutable mit Immutable

In [75]:
list1 = [(1, 2, 3), (4, 5, 6)]

# Erstelle eine Liste (Mutable) mit zwei Tupeln (Immutable) als Elemente

In [76]:
print(list1)

[(1, 2, 3), (4, 5, 6)]


In [77]:
print(hex(id(list1)))

0x1f19a5c0d00


In [78]:
list1[0] = (8,9,10)
list1[1] = 10
print(list1)

# Schreibe beim ersten Element eine neue Tupel und beim zweiten den Wert 10 rein

[(8, 9, 10), 10]


In [79]:
print(hex(id(list1)))

0x1f19a5c0d00


# OS

In [80]:
# Aktuelles Arbeitsverzeichnis
os.getcwd()

'c:\\Users\\marce\\Google Drive\\Python\\Lernen'

## Pfadarten

In [81]:
os.listdir(".") # Relativer Pfad

['.git',
 '.gitignore',
 '.ipynb_checkpoints',
 '.vscode',
 'Allgemein_Info.ipynb',
 'dask-worker-space',
 'Dask_Infos.ipynb',
 'data',
 'Grundsteuer.ipynb',
 'Imports.ipynb',
 'NBA_DataAnalysis.ipynb',
 'Netflix_DataAnalysis.ipynb',
 'Numpy_Infos.ipynb',
 'Pandas_Infos.ipynb',
 'Plotly_Infos.ipynb',
 'README.md',
 'Seaborn_Infos.ipynb',
 'test.ipynb',
 '_Git_Befehle.txt']

In [82]:
os.listdir("/Users/marce/Google Drive/Python/Lernen") # Absoluter Pfad

['.git',
 '.gitignore',
 '.ipynb_checkpoints',
 '.vscode',
 'Allgemein_Info.ipynb',
 'dask-worker-space',
 'Dask_Infos.ipynb',
 'data',
 'Grundsteuer.ipynb',
 'Imports.ipynb',
 'NBA_DataAnalysis.ipynb',
 'Netflix_DataAnalysis.ipynb',
 'Numpy_Infos.ipynb',
 'Pandas_Infos.ipynb',
 'Plotly_Infos.ipynb',
 'README.md',
 'Seaborn_Infos.ipynb',
 'test.ipynb',
 '_Git_Befehle.txt']

## Verzeichnis erstellen

In [83]:
os.makedirs("./data/allgemein", exist_ok=True)

In [84]:
"allgemein" in os.listdir("./data") # Wurde das Verzeichnis erstellt ?

True

In [85]:
url1 = 'https://gist.githubusercontent.com/aakashns/257f6e6c8719c17d0e498ea287d1a386/raw/7def9ef4234ddf0bc82f855ad67dac8b971852ef/loans1.txt'
url2 = 'https://gist.githubusercontent.com/aakashns/257f6e6c8719c17d0e498ea287d1a386/raw/7def9ef4234ddf0bc82f855ad67dac8b971852ef/loans2.txt'
url3 = 'https://gist.githubusercontent.com/aakashns/257f6e6c8719c17d0e498ea287d1a386/raw/7def9ef4234ddf0bc82f855ad67dac8b971852ef/loans3.txt'

 Die heruntergeladenen Dateien in das neu erstellte Verzeichnis laden

In [ ]:
urlretrieve(url1, './data/allgemein/loans1.txt')

('./data/allgemein/loans1.txt', <http.client.HTTPMessage at 0x18549487cd0>)

In [ ]:
urlretrieve(url2, './data/allgemein/loans2.txt')

('./data/allgemein/loans2.txt', <http.client.HTTPMessage at 0x18549701fa0>)

In [ ]:
urlretrieve(url3, './data/allgemein/loans3.txt')

('./data/allgemein/loans3.txt', <http.client.HTTPMessage at 0x1854c2a7e20>)

In [86]:
os.listdir("./data/allgemein/")

['loans1.txt', 'loans2.txt', 'loans3.txt']

## Lesen von einer Datei

In [87]:
file1 = open('./data/allgemein/loans1.txt', mode='r')

In [88]:
file1_contents = file1.read()

The first line of the file is the header, indicating what each of the numbers on the remaining lines represents. Each of the remaining lines provides information about a loan. Thus, the second line 10000,36,0.08,20000 represents a loan with:

- an amount of <b>10000 Dollar</b>
- duration of <b>36 months</b>
- rate of interest of <b>8 percentage per annum</b> and
- a down payment of <b>20000 Dollar</b>

In [89]:
print(file1_contents)

amount,duration,rate,down_payment
100000,36,0.08,20000
200000,12,0.1,
628400,120,0.12,100000
4637400,240,0.06,
42900,90,0.07,8900
916000,16,0.13,
45230,48,0.08,4300
991360,99,0.08,
423000,27,0.09,47200


In [90]:
file1.close()

In [92]:
file1.read() 

# Nach dem eine Datei geschlossen wurde, kann sie nicht mehr benutzt werden.

ValueError: I/O operation on closed file.

## Dateien automatisch lesen und schließen mit (with) (ContextManager)

In [93]:
with open('./data/allgemein/loans2.txt') as file2:
    file2_contents = file2.read()
    print(file2_contents)

amount,duration,rate,down_payment
828400,120,0.11,100000
4633400,240,0.06,
42900,90,0.08,8900
983000,16,0.14,
15230,48,0.07,4300


In [94]:
file2.read()

ValueError: I/O operation on closed file.

## Datei Zeile für Zeile lesen

In [95]:
with open('./data/allgemein/loans3.txt', 'r') as file3:
    file3_lines = file3.readlines()

In [96]:
file3_lines

['amount,duration,rate,down_payment\n',
 '45230,48,0.07,4300\n',
 '883000,16,0.14,\n',
 '100000,12,0.1,\n',
 '728400,120,0.12,100000\n',
 '3637400,240,0.06,\n',
 '82900,90,0.07,8900\n',
 '316000,16,0.13,\n',
 '15230,48,0.08,4300\n',
 '991360,99,0.08,\n',
 '323000,27,0.09,4720010000,36,0.08,20000\n',
 '528400,120,0.11,100000\n',
 '8633400,240,0.06,\n',
 '12900,90,0.08,8900']

## Verarbeitung von Daten aus Dateien

In [97]:
def parse_headers(header_line):
    return header_line.strip().split(',')

Die strip-Methode entfernt alle zusätzlichen Leerzeichen und das Zeilenumbruchzeichen \n. Die Split-Methode zerlegt eine Zeichenkette in eine Liste unter Verwendung des angegebenen Trennzeichens (, in diesem Fall).

In [98]:
file3_lines[0]

'amount,duration,rate,down_payment\n'

In [99]:
headers = parse_headers(file3_lines[0])

In [100]:
headers

['amount', 'duration', 'rate', 'down_payment']

In [101]:
def parse_values(data_line):
    values = []
    for item in data_line.strip().split(','):
        values.append(float(item))
    return values

In [102]:
file3_lines[1]

'45230,48,0.07,4300\n'

In [103]:
parse_values(file3_lines[1])

[45230.0, 48.0, 0.07, 4300.0]

In [104]:
file3_lines[2]

'883000,16,0.14,\n'

In [105]:
parse_values(file3_lines[2])

# ValueError, weil '' nicht in einen Float umgewandelt werden kann. Die Funktion parse_values kann erweitert werden, um diesen Ausnahmefall zu behandeln.

ValueError: could not convert string to float: ''

In [106]:
def parse_values(data_line):
    values = []
    for item in data_line.strip().split(','):
        if item == '':
            values.append(0.0)
        else:
            try:
                values.append(float(item))
            except ValueError:
                values.append(item)
    return values

In [107]:
file3_lines[2]

'883000,16,0.14,\n'

In [108]:
parse_values(file3_lines[2])

[883000.0, 16.0, 0.14, 0.0]

# Itertools

- Modul von der Standardbibliothek<br>
- Gibt einen Iterator als Returnwert zurück

In [2]:
import itertools

## Infinite Itertools

Bei diesen Iteratoren muss man selbst bestimmen wann sie aufhören sollen (Breakpoint festlegen). Ansonsten geht, wie der Name schon andeutet, es unendlich fortgeführt wird.

### count

In [2]:
from itertools import count

In [3]:
def iter_count(start, step):
    counter = count(start=start, step=step)
    
    for c in counter:
        print(c)
        
        if c == 50:
            break
        
iter_count(20, 5)

20
25
30
35
40
45
50


### cycle

In [ ]:
from itertools import cycle

In [ ]:
def iter_cycle(ele):
    i = 0
    cycler = cycle(ele)
    while i < 8:
        print(next(cycler), end=" ")
        i += 1

iter_cycle("HELLO")

H E L L O H E L 

### repeat

In [5]:
from itertools import repeat

In [9]:
def iter_repeat(ele, max_rep):
    repater = repeat(ele, max_rep)
    
    for x in repater:
        print(x)

iter_repeat("This is a test", 6)

This is a test
This is a test
This is a test
This is a test
This is a test
This is a test


## Terminating Iterators

### accumulate

In [14]:
from itertools import accumulate

In [17]:
a = [1,2,3,4]
acc = accumulate(a)
print(a)
print(list(acc))
# 1+2 = 3, 3+3 = 6, 6+4 = 10

[1, 2, 3, 4]
[1, 3, 6, 10]


In [ ]:
import operator

a = [1,2,3,4]
acc = accumulate(a, func=operator.mul) # Multiplikation
print(a)
print(list(acc))

[1, 2, 3, 4]
[1, 2, 6, 24]


In [ ]:
a = [1,2,55,3,4]
acc = accumulate(a, func=max) # Maximaler Wert
print(a)
print(list(acc))

[1, 2, 55, 3, 4]
[1, 2, 55, 55, 55]


In [ ]:
a = [1,2,55,3,4]
acc = accumulate(a, func=min) # Minimaler Wert
print(a)
print(list(acc))

[1, 2, 55, 3, 4]
[1, 1, 1, 1, 1]


### chain

Bessere Performanz im Vergleich zu "concatenate" von zwei Listen. "chain" ist keine Verkettung von zwei Listen. Es kommt ein Iterator zurück, der es einem ermöglicht, jeweils ein Element abzurufen und dieses nach Bedarf zu verarbeiten und zu verwenden. Dementsprechend speichert man nicht alles im Speicher 

In [18]:
from itertools import chain

In [19]:
def iter_chain(ele1, ele2):
    chainer = chain(ele1, ele2)
    print(list(chainer))
    
iter_chain("HEL", "LO")

['H', 'E', 'L', 'L', 'O']


In [25]:
def iter_chain_from_iterable(iterable):
    chainer = chain.from_iterable(iterable)
    print(list(chainer))
    
iter_chain_from_iterable([[1,2,3],[4,5,6],[7,8,9]])
# Eine verzweigte Liste kann damit "flach" gemacht werden. 

[1, 2, 3, 4, 5, 6, 7, 8, 9]


### compress

In [26]:
from itertools import compress

In [30]:
def iter_compress(data, select):
    compressor = compress(data=data, selectors=select)
    print(list(compressor))
    
iter_compress([["a","b","c"], [1,2,3], {"a":1, "b":2, "c":3}],
              [True,False,1])

# True / 1 = ["a","b","c"] (Element wird behalten)
# False / 0 = [1,2,3] (Element wird nicht behalten)

[['a', 'b', 'c'], {'a': 1, 'b': 2, 'c': 3}]


### groupby

In [69]:
from itertools import groupby

In [73]:
def iter_groupby(data, lambdafunc):
    groupby_func = groupby(data, key=lambdafunc)
    for k, v in groupby_func:
        print(k, list(v))
        
iter_groupby([1,33,2,55,3,4], lambda x: x>=3)

False [1]
True [33]
False [2]
True [55, 3, 4]


In [129]:
individuals = [
    {
        'name': 'Olaf Scholz',
        'city': 'Osnabrück',
        'state': 'Niedersachsen'
    },
    {
        'name': 'Angela Merkel',
        'city': 'Hamburg',
        'state': 'Hamburg'
    },
    {
        'name': 'Otto Walkes',
        'city': 'Emden',
        'state': 'Niedersachsen'
    }, 
    {
        'name': 'Gerhard Schröder',
        'city': 'Mossenberg-Woehren',
        'state': 'Nordrhein-Westfalen'
    },
    {
        'name': 'Per Mertesacker',
        'city': 'Hannover',
        'state': 'Niedersachsen'
    }
]

def get_state(individuals):
    return individuals['state']

individuals = sorted(individuals, key=get_state)

individuals_groupby = groupby(individuals, key=get_state)

for state, item in individuals_groupby:
    print(state)
    print(list(item), end="\n\n")
    # for chancellor in group:
    #     print(chancellor)
    # print()

Hamburg
[{'name': 'Angela Merkel', 'city': 'Hamburg', 'state': 'Hamburg'}]

Niedersachsen
[{'name': 'Olaf Scholz', 'city': 'Osnabrück', 'state': 'Niedersachsen'}, {'name': 'Otto Walkes', 'city': 'Emden', 'state': 'Niedersachsen'}, {'name': 'Per Mertesacker', 'city': 'Hannover', 'state': 'Niedersachsen'}]

Nordrhein-Westfalen
[{'name': 'Gerhard Schröder', 'city': 'Mossenberg-Woehren', 'state': 'Nordrhein-Westfalen'}]



### islice

In [74]:
from itertools import islice

In [88]:
def iter_islice(data, start, stop, step):
    islicer = islice(data, start, stop, step)
    for item in islicer:
        print(item)

iter_islice([1,2,3,4,5,6,7], 2, 8, 2)

3
5
7


Wird verwendet, wenn beispielsweise ein Iterator zu Groß für den Speicher ist und man nur ein paar Zeilen möchte. <br> (siehe Datei 'data/allgemein/test.log')

In [94]:
with open('data/allgemein/test.log', 'r') as f:
    header = islice(f, 3) # Ersten drei Zeilen ausgeben
    
    for line in header:
        print(line, end="")
        

Date: 2022-11-08
Author: Marcel
Description: Das ist ein Beispiel


### pairwise

In [31]:
from itertools import pairwise

In [33]:
def iter_pairwise(iterable):
    pairwiser = pairwise(iterable)
    print(list(pairwiser))
    
iter_pairwise([1,2,3,4,5,6,7,8,9,10])

[(1, 2), (2, 3), (3, 4), (4, 5), (5, 6), (6, 7), (7, 8), (8, 9), (9, 10)]


In [1]:
from itertools import product, permutations, combinations, accumulate, groupby

## Combinatoric Iterators

### product

In [34]:
from itertools import product

In [36]:
def iter_product(iter1, iter2):
    res_product = product(iter1, iter2)
    print(list(res_product))

iter_product([1,2],[2,3])

[(1, 2), (1, 3), (2, 2), (2, 3)]


In [43]:
def iter_product_repeat(iter1, iter2):
    res_product_rep = product(iter1, iter2, repeat=2)
    print(list(res_product_rep))

iter_product_repeat([1,2],[3])

[(1, 3, 1, 3), (1, 3, 2, 3), (2, 3, 1, 3), (2, 3, 2, 3)]


### permutations

In [44]:
from itertools import permutations

In [49]:
# Alle Permutationsmöglichkeiten

def iter_permutations(iter):
    res_permutations = permutations(iter)
    print(list(res_permutations))
    
iter_permutations([1,2,3,4])

[(1, 2, 3, 4), (1, 2, 4, 3), (1, 3, 2, 4), (1, 3, 4, 2), (1, 4, 2, 3), (1, 4, 3, 2), (2, 1, 3, 4), (2, 1, 4, 3), (2, 3, 1, 4), (2, 3, 4, 1), (2, 4, 1, 3), (2, 4, 3, 1), (3, 1, 2, 4), (3, 1, 4, 2), (3, 2, 1, 4), (3, 2, 4, 1), (3, 4, 1, 2), (3, 4, 2, 1), (4, 1, 2, 3), (4, 1, 3, 2), (4, 2, 1, 3), (4, 2, 3, 1), (4, 3, 1, 2), (4, 3, 2, 1)]


In [48]:
def iter_permutations_limited(iter):
    res_permutations_limited = permutations(iter, r=2)
    print(list(res_permutations_limited))
    
iter_permutations_limited([1,2,3,4])

# r gibt an wie viel Elemente pro Eintrag verwendet werden sollen

[(1, 2), (1, 3), (1, 4), (2, 1), (2, 3), (2, 4), (3, 1), (3, 2), (3, 4), (4, 1), (4, 2), (4, 3)]


### combinations

In [51]:
from itertools import combinations

In [65]:
# Einzigartige Kombinationen
def iter_combinations(iter, size):
    res_combinations = combinations(iter, size)
    print(f"Size {size}: {list(res_combinations)}")
    
list_comb = [1,2,3]
iter_combinations(list_comb, 1)
iter_combinations(list_comb, 2)
iter_combinations(list_comb, 3)


Size 1: [(1,), (2,), (3,)]
Size 2: [(1, 2), (1, 3), (2, 3)]
Size 3: [(1, 2, 3)]


In [60]:
from itertools import combinations_with_replacement

In [67]:
# Kombinationen mit zurücklegen

def iter_combinations_replace(iter, size):
    res_combinations_replace = combinations_with_replacement(iter, size)
    print(f"Size {size}: {list(res_combinations_replace)}")
    
list_comb = [1,2,3]
iter_combinations_replace(list_comb, 1)
iter_combinations_replace(list_comb, 2)
iter_combinations_replace(list_comb, 3)

Size 1: [(1,), (2,), (3,)]
Size 2: [(1, 1), (1, 2), (1, 3), (2, 2), (2, 3), (3, 3)]
Size 3: [(1, 1, 1), (1, 1, 2), (1, 1, 3), (1, 2, 2), (1, 2, 3), (1, 3, 3), (2, 2, 2), (2, 2, 3), (2, 3, 3), (3, 3, 3)]
